File: frontend/src/api/podcasts.ts
import { apiRequest, ApiError } from "./client";
import { API_PREFIX } from "./config";
import type { QueueTrack } from "@/player/types";

/**
 * Local response types for the podcasts API. ``types.ts`` is regenerated
 * from a running server (``npm run api:gen``); until then these interfaces
 * mirror the backend ``PodcastResponse``/``PodcastEpisodeResponse`` schemas.
 */
export interface PodcastResponse {
  id: string;
  feed_url: string;
  title: string;
  description: string | null;
  author: string | null;
  link: string | null;
  image_url: string | null;
  language: string | null;
  categories: string[];
  explicit: boolean;
  episode_count: number;
  unplayed_count: number;
  latest_episode_at: string | null;
  last_fetched_at: string | null;
  last_error: string | null;
  following: boolean;
}

export interface PodcastEpisodeResponse {
  id: string;
  podcast_id: string;
  guid: string;
  title: string;
  description: string | null;
  link: string | null;
  image_url: string | null;
  audio_url: string;
  audio_type: string | null;
  audio_length: number | null;
  duration_seconds: number | null;
  published_at: string | null;
  season_number: number | null;
  episode_number: number | null;
  episode_type: string | null;
  played: boolean;
}

export interface OpmlImportResponse {
  subscribed: number;
  skipped: number;
  failed: number;
  errors: string[];
}

export type PodcastSyncMode = "pull" | "bidirectional";
export type PodcastSyncServerType = "gpodder" | "nextcloud";

export interface PodcastSyncConfigResponse {
  server_type: PodcastSyncServerType;
  server_url: string;
  username: string;
  device_id: string;
  mode: PodcastSyncMode;
  enabled: boolean;
  has_password: boolean;
  last_synced_at: string | null;
  last_error: string | null;
}

export interface PodcastSyncConfigRequest {
  server_type?: PodcastSyncServerType;
  server_url: string;
  username: string;
  /** Omitted keeps the stored password; "" clears it. */
  password?: string;
  device_id?: string;
  mode?: PodcastSyncMode;
  enabled?: boolean;
}

export interface PodcastSyncResultResponse {
  subscribed: number;
  unsubscribed: number;
  pushed_adds: number;
  pushed_removes: number;
  errors: string[];
}

export type PodcastSortBy = "latest" | "episodes" | "unplayed" | "name";

export function listPodcasts(params?: {
  q?: string;
  limit?: number;
  offset?: number;
  sort_by?: PodcastSortBy | string;
  sort_dir?: "asc" | "desc";
}): Promise<PodcastResponse[]> {
  return apiRequest<PodcastResponse[]>("/podcasts/", { query: params });
}

export function followPodcast(feedUrl: string): Promise<PodcastResponse> {
  return apiRequest<PodcastResponse>("/podcasts/", {
    method: "POST",
    body: { feed_url: feedUrl },
  });
}

export function getPodcast(podcastId: string): Promise<PodcastResponse> {
  return apiRequest<PodcastResponse>(`/podcasts/${podcastId}`);
}

export function unfollowPodcast(podcastId: string): Promise<void> {
  return apiRequest<void>(`/podcasts/${podcastId}`, { method: "DELETE" });
}

export function refreshPodcast(podcastId: string): Promise<PodcastResponse> {
  return apiRequest<PodcastResponse>(`/podcasts/${podcastId}/refresh`, {
    method: "POST",
  });
}

export function listEpisodes(
  podcastId: string,
  params?: { limit?: number; offset?: number; sort?: "newest" | "oldest" },
): Promise<PodcastEpisodeResponse[]> {
  return apiRequest<PodcastEpisodeResponse[]>(
    `/podcasts/${podcastId}/episodes`,
    { query: params },
  );
}

export function markEpisodePlayed(episodeId: string): Promise<void> {
  return apiRequest<void>(`/podcasts/episodes/${episodeId}/played`, {
    method: "POST",
  });
}

export function markEpisodeUnplayed(episodeId: string): Promise<void> {
  return apiRequest<void>(`/podcasts/episodes/${episodeId}/played`, {
    method: "DELETE",
  });
}

/**
 * OPML export is a file download, not a JSON payload — expose the URL so
 * the view can hand it to an anchor and let the browser stream it.
 */
export function opmlExportUrl(): string {
  return `${API_PREFIX}/podcasts/opml`;
}

export function importOpml(file: File): Promise<OpmlImportResponse> {
  const body = new FormData();
  body.append("file", file);
  return apiRequest<OpmlImportResponse>("/podcasts/opml/import", {
    method: "POST",
    body,
  });
}

/**
 * Fetch the caller's GPodder sync configuration. Returns ``null`` when no
 * sync is configured (the API answers 404).
 */
export async function getPodcastSyncConfig(): Promise<PodcastSyncConfigResponse | null> {
  try {
    return await apiRequest<PodcastSyncConfigResponse>("/podcasts/sync");
  } catch (err) {
    if (err instanceof ApiError && err.status === 404) {
      return null;
    }
    throw err;
  }
}

export function putPodcastSyncConfig(
  body: PodcastSyncConfigRequest,
): Promise<PodcastSyncConfigResponse> {
  return apiRequest<PodcastSyncConfigResponse>("/podcasts/sync", {
    method: "PUT",
    body,
  });
}

export function deletePodcastSyncConfig(): Promise<void> {
  return apiRequest<void>("/podcasts/sync", { method: "DELETE" });
}

export function syncPodcastsNow(): Promise<PodcastSyncResultResponse> {
  return apiRequest<PodcastSyncResultResponse>("/podcasts/sync/now", {
    method: "POST",
  });
}

/**
 * Build a ``QueueTrack`` for a podcast episode. Episodes have no local
 * track row — the player streams ``audio_url`` directly via
 * ``stream_url`` and ``remote`` skips library links/history reporting.
 */
export function episodeToQueueTrack(
  episode: PodcastEpisodeResponse,
  podcast: PodcastResponse,
): QueueTrack {
  return {
    id: episode.id,
    title: episode.title,
    artist_id: "",
    artist_name: podcast.author ?? podcast.title,
    album_title: podcast.title,
    artwork_url: episode.image_url ?? podcast.image_url ?? undefined,
    duration: episode.duration_seconds ?? undefined,
    visibility: "public",
    tags: [],
    genres: [],
    is_external: false,
    stream_url: episode.audio_url,
    remote: true,
    remote_url: episode.link ?? undefined,
    podcast_episode_id: episode.id,
    podcast_id: episode.podcast_id,
    in_collection: false,
  };
}

File: songhive/api/routes/podcasts.py
"""
Podcast routes — RSS/Atom feed subscriptions under ``/api/v1/podcasts``.

Podcasts are followed by RSS link: ``POST /`` fetches and parses the feed,
stores show and episode metadata, and records a per-user subscription.
Episode audio is streamed from the source enclosure URL, never copied into
local storage. ``GET /opml`` exports the caller's subscriptions and
``POST /opml/import`` follows every feed in an uploaded OPML document.
"""

import json
import logging
from datetime import datetime
from typing import List, Optional

from fastapi import APIRouter, Depends, File, HTTPException, Query, Response, UploadFile, status
from pydantic import BaseModel, ConfigDict
from sqlalchemy.ext.asyncio import AsyncSession

from ...config.schema import SonghiveConfig
from ...models.podcast import Podcast, PodcastEpisode
from ...models.user import User
from ...services import podcasts as podcasts_service
from ...services.podcasts import FeedFetchError, PodcastStats
from .._common import Pagination, get_pagination
from ..deps import get_config, get_current_user, get_db

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/podcasts")

OPML_MEDIA_TYPE = "text/x-opml"


class PodcastResponse(BaseModel):
    """Public podcast response."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    feed_url: str
    title: str
    description: Optional[str] = None
    author: Optional[str] = None
    link: Optional[str] = None
    image_url: Optional[str] = None
    language: Optional[str] = None
    categories: List[str] = []
    explicit: bool = False
    episode_count: int = 0
    unplayed_count: int = 0
    latest_episode_at: Optional[datetime] = None
    last_fetched_at: Optional[datetime] = None
    last_error: Optional[str] = None
    following: bool = False


class PodcastEpisodeResponse(BaseModel):
    """A single podcast episode; ``audio_url`` is the remote enclosure."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    podcast_id: str
    guid: str
    title: str
    description: Optional[str] = None
    link: Optional[str] = None
    image_url: Optional[str] = None
    audio_url: str
    audio_type: Optional[str] = None
    audio_length: Optional[int] = None
    duration_seconds: Optional[int] = None
    published_at: Optional[datetime] = None
    season_number: Optional[int] = None
    episode_number: Optional[int] = None
    episode_type: Optional[str] = None
    played: bool = False


class PodcastFollowRequest(BaseModel):
    """Follow a podcast by RSS/Atom feed URL."""

    feed_url: str


class OpmlImportResponse(BaseModel):
    """Summary of an OPML subscription import."""

    subscribed: int = 0
    skipped: int = 0
    failed: int = 0
    errors: List[str] = []


def _check_podcasts(config: SonghiveConfig) -> None:
    """Raise 404 when the podcasts feature is disabled."""
    if not config.podcasts.enabled:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")


def _categories(podcast: Podcast) -> List[str]:
    """Decode the JSON category list stored on the row."""
    if not podcast.categories:
        return []
    try:
        value = json.loads(podcast.categories)
    except (TypeError, ValueError):
        return []
    return [str(item) for item in value] if isinstance(value, list) else []


def _podcast_response(podcast: Podcast, following: bool, stats: PodcastStats) -> PodcastResponse:
    return PodcastResponse(
        id=str(podcast.id),
        feed_url=podcast.feed_url,
        title=podcast.title,
        description=podcast.description,
        author=podcast.author,
        link=podcast.link,
        image_url=podcast.image_url,
        language=podcast.language,
        categories=_categories(podcast),
        explicit=podcast.explicit,
        episode_count=stats.episode_count,
        unplayed_count=stats.unplayed_count,
        latest_episode_at=stats.latest_episode_at,
        last_fetched_at=podcast.last_fetched_at,
        last_error=podcast.last_error,
        following=following,
    )


def _episode_response(episode: PodcastEpisode, played: bool = False) -> PodcastEpisodeResponse:
    return PodcastEpisodeResponse(
        id=str(episode.id),
        podcast_id=str(episode.podcast_id),
        guid=episode.guid,
        title=episode.title,
        description=episode.description,
        link=episode.link,
        image_url=episode.image_url,
        audio_url=episode.audio_url,
        audio_type=episode.audio_type,
        audio_length=episode.audio_length,
        duration_seconds=episode.duration_seconds,
        published_at=episode.published_at,
        season_number=episode.season_number,
        episode_number=episode.episode_number,
        episode_type=episode.episode_type,
        played=played,
    )


async def _get_podcast_or_404(db: AsyncSession, podcast_id: str) -> Podcast:
    podcast = await podcasts_service.get_podcast(db, podcast_id)
    if podcast is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Podcast not found")
    return podcast


@router.get("/", response_model=List[PodcastResponse])
async def list_podcasts(
    response: Response,
    q: Optional[str] = Query(None, max_length=200),
    sort_by: str = Query("latest", pattern="^(latest|episodes|unplayed|name)$"),
    sort_dir: Optional[str] = Query(None, pattern="^(asc|desc)$"),
    current_user: User = Depends(get_current_user),
    pagination: Pagination = Depends(get_pagination),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """
    List podcasts the current user follows.

    ``sort_by`` accepts ``latest`` (newest episode first, the default),
    ``episodes``, ``unplayed`` or ``name``; ``q`` filters on title, author
    and description.
    """
    _check_podcasts(config)
    entries, total = await podcasts_service.list_subscribed_podcasts(
        db,
        current_user,
        limit=pagination.limit,
        offset=pagination.offset,
        search=q,
        sort_by=sort_by,
        sort_dir=sort_dir,
    )
    pagination.set_total(response, total)
    return [_podcast_response(entry.podcast, True, entry.stats) for entry in entries]


@router.post("/", response_model=PodcastResponse, status_code=201)
async def follow_podcast(
    body: PodcastFollowRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """Follow a podcast by RSS/Atom feed URL."""
    _check_podcasts(config)
    try:
        podcast, _ = await podcasts_service.subscribe(db, current_user, body.feed_url, config)
    except FeedFetchError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Could not fetch or parse feed: {exc}",
        ) from exc
    stats = await podcasts_service.podcast_stats(db, current_user, [str(podcast.id)])
    return _podcast_response(podcast, True, stats.get(str(podcast.id), PodcastStats()))


@router.get("/opml")
async def export_opml(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """Export the current user's podcast subscriptions as an OPML document."""
    _check_podcasts(config)
    entries, _ = await podcasts_service.list_subscribed_podcasts(db, current_user, limit=1000, sort_by="name")
    username = current_user.display_name or current_user.username
    document = podcasts_service.render_opml(f"{username}'s podcasts", [entry.podcast for entry in entries])
    return Response(
        content=document,
        media_type=OPML_MEDIA_TYPE,
        headers={"Content-Disposition": 'attachment; filename="podcasts.opml"'},
    )


@router.post("/opml/import", response_model=OpmlImportResponse)
async def import_opml(
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """
    Follow every feed listed in an uploaded OPML document.

    Each feed is fetched and parsed synchronously; feeds that fail are
    reported in ``errors`` without aborting the rest of the import.
    """
    _check_podcasts(config)
    data = await file.read()
    try:
        entries = podcasts_service.parse_opml(data)
    except FeedFetchError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Could not parse OPML document: {exc}",
        ) from exc

    result = OpmlImportResponse()
    for feed_url, _ in entries:
        try:
            __, created = await podcasts_service.subscribe(db, current_user, feed_url, config)
        except FeedFetchError as exc:
            result.failed += 1
            result.errors.append(f"{feed_url}: {exc}")
            continue
        if created:
            result.subscribed += 1
        else:
            result.skipped += 1
    return result


@router.get("/{podcast_id}", response_model=PodcastResponse)
async def get_podcast(
    podcast_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """Get a podcast by id."""
    _check_podcasts(config)
    podcast = await _get_podcast_or_404(db, podcast_id)
    following = await podcasts_service.is_subscribed(db, current_user, podcast_id)
    stats = await podcasts_service.podcast_stats(db, current_user, [podcast_id])
    return _podcast_response(podcast, following, stats.get(podcast_id, PodcastStats()))


@router.delete("/{podcast_id}", status_code=204)
async def unfollow_podcast(
    podcast_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """Unfollow a podcast (drops the subscription, not the shared feed row)."""
    _check_podcasts(config)
    removed = await podcasts_service.unsubscribe(db, current_user, podcast_id)
    if not removed:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Subscription not found")


@router.post("/{podcast_id}/refresh", response_model=PodcastResponse)
async def refresh_podcast(
    podcast_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """Re-fetch and reparse the podcast's feed right now."""
    _check_podcasts(config)
    podcast = await _get_podcast_or_404(db, podcast_id)
    try:
        await podcasts_service.refresh_podcast(db, podcast, config, force=True)
    except FeedFetchError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Could not refresh feed: {exc}",
        ) from exc
    following = await podcasts_service.is_subscribed(db, current_user, podcast_id)
    stats = await podcasts_service.podcast_stats(db, current_user, [podcast_id])
    return _podcast_response(podcast, following, stats.get(podcast_id, PodcastStats()))


@router.get("/{podcast_id}/episodes", response_model=List[PodcastEpisodeResponse])
async def list_episodes(
    podcast_id: str,
    response: Response,
    sort: str = Query("newest", pattern="^(newest|oldest)$"),
    current_user: User = Depends(get_current_user),
    pagination: Pagination = Depends(get_pagination),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """List a podcast's episodes, newest first."""
    _check_podcasts(config)
    await _get_podcast_or_404(db, podcast_id)
    episodes, total = await podcasts_service.list_episodes(
        db, podcast_id, limit=pagination.limit, offset=pagination.offset, oldest_first=sort == "oldest"
    )
    pagination.set_total(response, total)
    played = await podcasts_service.played_episode_ids(db, current_user, [str(e.id) for e in episodes])
    return [_episode_response(episode, str(episode.id) in played) for episode in episodes]


@router.get("/episodes/{episode_id}", response_model=PodcastEpisodeResponse)
async def get_episode(
    episode_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """Get a single podcast episode by id."""
    _check_podcasts(config)
    episode = await podcasts_service.get_episode(db, episode_id)
    if episode is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Episode not found")
    played = await podcasts_service.played_episode_ids(db, current_user, [str(episode.id)])
    return _episode_response(episode, str(episode.id) in played)


@router.post("/episodes/{episode_id}/played", status_code=204)
async def mark_episode_played(
    episode_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """Mark an episode as played for the current user (idempotent)."""
    _check_podcasts(config)
    episode = await podcasts_service.mark_episode_played(db, current_user, episode_id)
    if episode is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Episode not found")


@router.delete("/episodes/{episode_id}/played", status_code=204)
async def mark_episode_unplayed(
    episode_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    config: SonghiveConfig = Depends(get_config),
):
    """Clear the played mark for an episode."""
    _check_podcasts(config)
    episode = await podcasts_service.mark_episode_unplayed(db, current_user, episode_id)
    if episode is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Episode not found")
